// camper/camper/project.js
(function() {
	var project;
	
	project = (function() {
		
		function project(id, client) {
			this.id = id;
			this.client = client;
		}

		project.prototype.info = function(cb) {
			return this.client.get('/projects/'+this.id, function(err, s, b) {
				if (err) return cb(err);
				if (s !== 200) {
					return cb(new Error('API Error: '+b));
				} else {
					return cb(null, b);
				}
			});
		};
		
		project.prototype.add = function(data, cb) {
			return this.client.post('/projects', data, function(err, s, b) {
				if (err) return cb(err);
				if (s !== 201) {
					return cb(new Error('API Error: '+b));
				} else {
					return cb(null, b);
				}
			});
		};

		project.prototype.update = function(data, cb) {
			return this.client.post('/projects/'+this.id, data, function(err, s, b) {
				if (err) return cb(err);
				if (s !== 201) {
					return cb(new Error('API Error: '+b));
				} else {
					return cb(null, b);
				}
			});
		};

		project.prototype.del = function(cb) {
			return this.client.del('/projects/'+this.id, function(err, s, b) {
				if (err) return cb(err);
				if (s !== 200) {
					return cb(new Error('API Error: '+b));
				} else {
					return cb(null, b);
				}
			});
		};

		return project;

	})();
	
	module.exports = project;
	
}).call(this);
